### PR TITLE
JSON opts

### PR DESCRIPTION
**Pull request by kalekale**
Adds support for the same encoding options as cheshire has.

### CODE REVIEW

File: src/clj/muuntaja/json.clj [hunk 1]
```diff
@@ -67,21 +69,25 @@
   "Create a Jackson Databind module to support Clojure datastructures.
 
   See make-mapper docstring for the documentation of the options."
-  [{:keys [keywordize? encoders]}]
+  [{:keys [keywordize? pretty? encoders date-format]}]
```
Comment (miikka) [inline src/clj/muuntaja/json.clj]: `pretty?` isn't used anywehere

File: src/clj/muuntaja/json.clj [hunk 2]
```diff
@@ -67,21 +69,25 @@
...
     (as-> module
         (doseq [[cls encoder-fn] encoders]
           (.addSerializer module cls (FunctionalSerializer. encoder-fn))))
     (cond->
-      ;; This key deserializer decodes the map keys into Clojure keywords.
-      keywordize? (.addKeyDeserializer Object (KeywordKeyDeserializer.)))))
+        ;; This key deserializer decodes the map keys into Clojure keywords.
+        keywordize? (.addKeyDeserializer Object (KeywordKeyDeserializer.)))))
+
+;mapper.enable(SerializationFeature.INDENT_OUTPUT)
```
Comment (miikka) [inline src/clj/muuntaja/json.clj]: ?

File: src/clj/muuntaja/json.clj [hunk 3]
```diff
@@ -99,12 +105,16 @@
   ([] (make-mapper {}))
   ([options]
    (doto (ObjectMapper.)
-     (.registerModule (make-clojure-module options)))))
+     (.registerModule (make-clojure-module options))
+     (cond-> (:pretty options) (.enable SerializationFeature/INDENT_OUTPUT)
+             (:escape-non-ascii options) (.enable ^"[Lcom.fasterxml.jackson.core.JsonGenerator$Feature;"
+                                                  (into-array [JsonGenerator$Feature/ESCAPE_NON_ASCII]))))))
 
+;(:escape-non-ascii options) (.enable JsonGenerator$Feature^ JsonGenerator$Feature/ESCAPE_NON_ASCII)
```
Comment (miikka) [inline src/clj/muuntaja/json.clj]: ?

File: src/clj/muuntaja/json.clj [hunk 4]
```diff
@@ -99,12 +105,16 @@
   ([] (make-mapper {}))
```
Comment (miikka) [inline src/clj/muuntaja/json.clj]: Please update this docstring to include the new keywords.

File: src/clj/muuntaja/json.clj [hunk 5]
```diff
@@ -116,13 +126,34 @@
      (.readValue mapper ^String data ^Class Object)
      (.readValue mapper ^InputStream data ^Class Object))))
 
+(defn from-json-with-opts
```
Comment (ikitommi) [inline src/clj/muuntaja/json.clj]: Could the `from-json` & `to-json` just check the argument type to see if it was given opts or mapper?

Comment (ikitommi) [inline src/clj/muuntaja/json.clj]: `muuntaja.core/create` works like this now.

File: src/java/muuntaja/jackson/DateSerializer.java [hunk 6]
```diff
@@ -11,12 +11,20 @@
...
+        formatter.setTimeZone(new SimpleTimeZone(0, "UTC"));
+    }
+
     public DateSerializer() {
         super(DateSerializer.class, true);
         formatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'");
         formatter.setTimeZone(new SimpleTimeZone(0, "UTC"));
     }
 
+
```
Comment (ikitommi) [inline src/java/muuntaja/jackson/DateSerializer.java]: space invaders?

File: src/java/muuntaja/jackson/DateSerializer.java [hunk 7]
```diff
@@ -11,12 +11,20 @@
...
     private final SimpleDateFormat formatter;
 
+    public DateSerializer(String dateFormat) {
+        super(DateSerializer.class, true);
+        formatter = new SimpleDateFormat(dateFormat);
+        formatter.setTimeZone(new SimpleTimeZone(0, "UTC"));
+    }
+
     public DateSerializer() {
         super(DateSerializer.class, true);
```
Comment (miikka) [inline src/java/muuntaja/jackson/DateSerializer.java]: This constructor could now just call the other constructor. `this("yyyy-MM-dd'T'HH:mm:ss'Z'");`

File: src/clj/muuntaja/json.clj [hunk 8]
```diff
@@ -118,38 +118,26 @@
 
   To configure, pass in an ObjectMapper created with make-mapper."
```
Comment (ikitommi) [inline src/clj/muuntaja/json.clj]: Docs are now out of sync

File: src/clj/muuntaja/json.clj [hunk 9]
```diff
@@ -99,29 +103,41 @@
...
   "Decode a value from a JSON string or InputStream.
 
   To configure, pass in an ObjectMapper created with make-mapper."
   ([data] (from-json data +default-mapper+))
-  ([data ^ObjectMapper mapper]
-   (if (string? data)
-     (.readValue mapper ^String data ^Class Object)
-     (.readValue mapper ^InputStream data ^Class Object))))
+  ([data arg]
+   (let [decode (fn [data ^ObjectMapper mapper]
```
Comment (ikitommi) [inline src/clj/muuntaja/json.clj]: the innerfunction gets generated in all function calls. Could be clearer (and faster) if the mapper is defined in a let used in a body `(let [mapper (cond...)] ....)`

File: src/clj/muuntaja/json.clj [hunk 10]
```diff
@@ -67,61 +69,78 @@
...
+        ;; This key deserializer decodes the map keys into Clojure keywords.
+        keywordize? (.addKeyDeserializer Object (KeywordKeyDeserializer.)))))
 
 (defn ^ObjectMapper make-mapper
   "Create an ObjectMapper with Clojure support.
 
   The optional first parameter is a map of options. The following options are
   available:
-
+  Encodin options:
```
Comment (miikka) [inline src/clj/muuntaja/json.clj]: [nit] `Encodin` -> `Encoding`

File: src/clj/muuntaja/json.clj [hunk 11]
```diff
@@ -67,61 +69,78 @@
...
+  :date-format      -- string for custom date formatting. If not set will use default \"yyyy-MM-dd'T'HH:mm:ss'Z'\"
   :encoders     --  a map of custom encoders where keys should be types and values
                     should be encoder functions
-  :keywordize?  --  set to true to convert map keys into keywords (default: false)
-
   Encoder functions take two parameters: the value to be encoded and a
   JsonGenerator object. The function should call JsonGenerator methods to emit
-  the desired JSON."
+  the desired JSON.
+  Decodin options:
```
Comment (miikka) [inline src/clj/muuntaja/json.clj]: [nit] `Decodin` -> `Decoding`.

A good idea to split the encoding and decoding options.

**Pull request merged**